pkg/upgrade/upgrade_state.go

```diff
@@ -0,0 +1,226 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upgrade
+
+import (
+	"context"
+	"strconv"
+
+	"github.com/go-logr/logr"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/Mellanox/network-operator/api/v1alpha1"
+	"github.com/Mellanox/network-operator/pkg/consts"
+)
+
+type NodeUpgradeState struct {
+	Node            *v1.Node
+	DriverPod       *v1.Pod
+	DriverDaemonSet *appsv1.DaemonSet
+}
+
+type ClusterUpgradeState struct {
+	Policy *v1alpha1.OfedUpgradePolicySpec
+
+	NodeStates map[string][]*NodeUpgradeState
+}
+
+func NewClusterUpgradeState() ClusterUpgradeState {
+	return ClusterUpgradeState{NodeStates: make(map[string][]*NodeUpgradeState)}
+}
+
+type ClusterUpgradeStateManager struct {
+	K8sClient         client.Client
+	Log               logr.Logger
+	DrainManager      DrainManager
+	PodRestartManager PodRestartManager
+}
+
+func NewClusterUpdateStateManager(
+	drainManager DrainManager,
+	podRestartManager PodRestartManager,
+	log logr.Logger,
+	k8sClient client.Client) *ClusterUpgradeStateManager {
+	manager := &ClusterUpgradeStateManager{
+		DrainManager:      drainManager,
+		PodRestartManager: podRestartManager,
+		Log:               log,
+		K8sClient:         k8sClient,
+	}
+
+	return manager
+}
+
+func (m *ClusterUpgradeStateManager) ApplyState(ctx context.Context, currentState *ClusterUpgradeState) {
+	m.Log.V(consts.LogLevelInfo).Info("State Manager, got state update")
+
+	policy := currentState.Policy
+	if !policy.AutoUpgrade {
+		m.Log.V(consts.LogLevelInfo).Info("Driver auto upgrade is disabled, skipping")
+		return
+	}
+
+	m.Log.V(consts.LogLevelInfo).Info("Node states:",
+		"Unknown", len(currentState.NodeStates[Unknown]),
+		Ready, len(currentState.NodeStates[Ready]),
+		UpgradeRequired, len(currentState.NodeStates[UpgradeRequired]),
+		Drain, len(currentState.NodeStates[Drain]),
+		DrainFailed, len(currentState.NodeStates[DrainFailed]),
+		PodRestart, len(currentState.NodeStates[PodRestart]))
+
+	upgradesInProgress := len(currentState.NodeStates[Drain]) +
+		len(currentState.NodeStates[PodRestart]) +
+		len(currentState.NodeStates[DrainFailed])
+
+	upgradesAvailable := policy.MaxParallelUpgrades - upgradesInProgress
+
+	m.Log.V(consts.LogLevelInfo).Info("Upgrades in progress",
+		"currently in progress", upgradesInProgress,
+		"max parallel upgrades", policy.MaxParallelUpgrades,
+		"upgrade slots available", upgradesAvailable)
+
+	// First, check if unknown or ready nodes need to be upgraded
+	m.ProcessReadyOrUnknownNodes(currentState, Unknown)
```

Comment:
Please, do not ignore errors in all Process handlers. I thinks we should rise all critical errors up to Reconcile loop and rely on controller-runtime error handling logic which will schedule retry for us.